File: backend-gin/internal/lib/utils/concurrency.go
package utils

import (
	"errors"
	"fmt"
	"sync"
)

func BatchProcessInChunks[T any, K any](
	batch []T, transform func(T) (K, error), chunkSize int,
) ([]K, error) {
	results := make([]K, len(batch))
	errs := make([]error, len(batch))

	for start := 0; start < len(batch); start += chunkSize {
		end := min(start+chunkSize, len(batch))
		chunk := batch[start:end]

		var wg sync.WaitGroup
		for i, item := range chunk {
			wg.Add(1)
			go func() {
				defer wg.Done()
				globalIdx := start + i
				processed, err := transform(item)
				results[globalIdx] = processed
				errs[globalIdx] = err
			}()
		}
		wg.Wait()
	}

	err := errors.Join(errs...)
	if err != nil {
		return results, fmt.Errorf("batch processing failed:\n%w", err)
	}

	return results, nil
}

func BatchProcessWithSemaphore[T any, K any](
	batch []T, transform func(T) (K, error), maxWorkers int,
) ([]K, error) {
	var wg sync.WaitGroup
	results := make([]K, len(batch))
	errs := make([]error, len(batch))
	usedWorkers := make(chan struct{}, maxWorkers)

	for i, item := range batch {
		wg.Add(1)
		go func() {
			defer wg.Done()
			usedWorkers <- struct{}{}
			defer func() { <-usedWorkers }()
			processed, err := transform(item)
			results[i] = processed
			errs[i] = err
		}()
	}

	wg.Wait()
	err := errors.Join(errs...)
	if err != nil {
		return results, fmt.Errorf("batch processing failed:\n%w", err)
	}

	return results, nil
}

File: backend-express/src/lib/utils/concurrency.ts
export async function processBatchByChunks<T, K>(
    batch: T[],
    transform: (item: T) => Promise<K>,
    chunkSize: number
): Promise<K[]> {
    const results: K[] = [];

    for (let i = 0; i < batch.length; i += chunkSize) {
        const chunk = batch.slice(i, i + chunkSize);
        const processed = await Promise.all(chunk.map(transform));
        results.push(...processed);
    }

    return results;
}

export async function processBatchWithSemaphore<T, K>(
    batch: T[],
    transform: (item: T) => Promise<K>,
    maxWorkers: number
): Promise<K[]> {
    const queue = batch.map((item, i) => ({ item, i }));
    const results: K[] = new Array(batch.length);

    async function worker() {
        while (queue.length) {
            const { item, i } = queue.shift()!;
            results[i] = await transform(item);
        }
    }

    await Promise.all(Array.from({ length: maxWorkers }, worker));
    return results;
}
